### PR TITLE
Add direct connect support

### PR DESCRIPTION
**Pull request by schirevko**

### CODE REVIEW

File: vkcs/dc/resource_bgp_neighbor.go [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+				Optional:    true,
+				Computed:    true,
+				Description: "Force set IP address of next-hop on BGP prefix to self even in iBGP",
+			},
+
+			"add_paths": schema.StringAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Activate BGP Add-Paths feature on peer.",
+				// sch: add validation
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_neighbor.go]: Added validation

File: vkcs/dc/resource_bgp_instance.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+	plan.CreatedAt = types.StringValue(bgpInstanceResp.CreatedAt)
+	plan.UpdatedAt = types.StringValue(bgpInstanceResp.UpdatedAt)
+
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *BGPInstanceResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var plan BGPInstanceResourceModel
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: I think we may keep "plan" for create and update, and use "state" for read and delete methods, according to naming used in hashicorp tutorials: https://developer.hashicorp.com/terraform/tutorials/providers-plugin-framework/providers-plugin-framework-resource-create

File: vkcs/dc/resource_bgp_instance.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+		region = r.config.GetRegion()
+	}
+
+	networkingClient, err := r.config.NetworkingV2Client(region, networking.SprutSDN)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating VKCS networking client", err.Error())
+		return
+	}
+
+	planID := plan.ID.ValueString()
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: Renamed to `id`

File: vkcs/internal/services/dc/v2/bgp_instances/requests.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"net/http"
+
+	"github.com/gophercloud/gophercloud"
+)
+
+type OptsBuilder interface {
+	Map() (map[string]interface{}, error)
+}
+type BGPInstanceCreate struct {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/internal/services/dc/v2/bgp_instances/requests.go:13]: Added

File: vkcs/internal/services/dc/v2/bgp_instances/results.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package bgp_instances
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+type BGPInstanceResp struct {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/internal/services/dc/v2/bgp_instances/results.go:7]: Renamed to `BGPInstance` etc

File: vkcs/internal/services/dc/v2/bgp_neighbors/urls.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package bgp_neighbors
+
+import "github.com/gophercloud/gophercloud"
+
+func baseURL() string {
+	return "direct_connect/dc_bgp_neighbors"
+}
+
+func bgpneighborsURL(c *gophercloud.ServiceClient) string {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/internal/services/dc/v2/bgp_neighbors/urls.go]: Renamed

File: vkcs/dc/resource_bgp_instance.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+
+func (r *BGPInstanceResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	r.config = req.ProviderData.(clients.Config)
+}
+
+func (r *BGPInstanceResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan BGPInstanceResourceModel
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: Used plan/data for update and data for everything else

File: vkcs/dc/resource_bgp_instance.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: Used plan/data for update and data for everything else

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,9 @@ description: |-
 ---
 
 # VKCS Provider's changelog
+#### v.0.4.3
```
[the remark this answers is not in the document]
Comment (schirevko) [reply CHANGELOG.md]: After rebase (unreleased) is present

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,9 +6,9 @@ description: |-
 ---
 
 # VKCS Provider's changelog
-
 #### v0.4.3 (unreleased)
 - Fix Kubernetes addon state refresh panic on error
+- Add Direct Connect resources and data sources
```
Comment (ftersin) [inline CHANGELOG.md:11]: place first and change 4.3 to 5.0

File: vkcs/dc/resource_router.go [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+				Description: "The availability zone in which to create the router. Changing this creates a new router",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+
+			"flavor": schema.StringAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Flavor of the router. Possible values can be obtained with vkcs_dc_api_options data source. Not to be confused with compute service flavors. Changing this creates a new router",
```
Comment (ftersin) [inline vkcs/dc/resource_router.go:64]: Lets mark "not to be confused..." as note.

File: examples/dc/bgp_instance/main.tf [hunk 11]
```diff
@@ -0,0 +1,10 @@
+resource "vkcs_dc_bgp_instance" "dc_bgp_instance" {
+    name = "tf-example"
+    description = "tf-example-description"
+    dc_router_id = vkcs_dc_router.dc_router.id
+    bgp_router_id = "192.168.1.2"
```
[the remark this answers is not in the document]
Comment (ftersin) [reply examples/dc/bgp_instance/main.tf:5]: does not it make sense to rename the argument?

File: vkcs/dc/resource_interface.go [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+			"bgp_announce_enabled": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Enable BGP announce of subnet attached to interface",
+			},
+
+			"ip_address": schema.StringAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "IP Address. Changing this creates a new resource",
```
Comment (ftersin) [inline vkcs/dc/resource_interface.go:107]: ip addres of what?

File: vkcs/dc/resource_interface.go [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+			},
+
+			"ip_netmask": schema.Int64Attribute{
+				Computed:    true,
+				Description: "IP Netmask",
+			},
+
+			"mac_address": schema.StringAttribute{
+				Computed:    true,
+				Description: "IP Netmask",
```
Comment (ftersin) [inline vkcs/dc/resource_interface.go:130]: MAC address?

File: vkcs/dc/resource_interface.go [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+			},
+
+			"port_id": schema.StringAttribute{
+				Computed:    true,
+				Description: "Port ID",
+			},
+
+			"sdn": schema.StringAttribute{
+				Computed:    true,
+				Description: "SDN",
```
Comment (ftersin) [inline vkcs/dc/resource_interface.go:120]: SDN of what?

File: examples/dc/bgp_instance/main.tf (hunk 11 above)
[the remark this answers is not in the document]
Comment (schirevko) [reply examples/dc/bgp_instance/main.tf:5]: Field is named according to backend, I think it's not great to have discrepancy between provider and service. Besides, field descriptions gives info that this is IP address: https://github.com/vk-cs/terraform-provider-vkcs/pull/282/files#diff-61fad4012c68179e341555a2aafd34ea992d27666872f5f562e158e8cc9414eaR80

File: CHANGELOG.md (hunk 9 above)
Comment (schirevko) [reply CHANGELOG.md:10]: Updated

File: vkcs/dc/resource_router.go (hunk 10 above)
Comment (schirevko) [reply vkcs/dc/resource_router.go]: Marked

File: vkcs/dc/resource_bgp_instance.go [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+			"ecmp_enabled": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Enable BGP ECMP behaviour on router",
+			},
+
+			"enabled": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Enable or disable item",
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: Added default values to descriptions

File: vkcs/dc/resource_interface.go (hunk 12 above)
Comment (schirevko) [reply vkcs/dc/resource_interface.go]: Expanded description

File: vkcs/dc/resource_interface.go (hunk 13 above)
Comment (schirevko) [reply vkcs/dc/resource_interface.go]: Fixed and expanded description

File: vkcs/dc/resource_interface.go (hunk 14 above)
Comment (schirevko) [reply vkcs/dc/resource_interface.go]: Expanded description

File: vkcs/dc/resource_bgp_instance.go [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+	}
+
+	bgpInstanceID := data.ID.ValueString()
+
+	bgpInstanceUpdateOpts := bgpinstances.UpdateOpts{
+		Name:        plan.Name.ValueString(),
+		Description: plan.Description.ValueString(),
+	}
+
+	if !plan.ECMPEnabled.IsUnknown() {
```
[the remark this answers is not in the document]
Comment (schirevko) [reply vkcs/dc/resource_bgp_instance.go]: Rewrote to use such function

**Pull request merged**